aiida/tools/visualization/graph.py

```diff
@@ -0,0 +1,740 @@
+
+        node_style["label"] = "\n".join(label)
+
+    if style_override is not None:
+        node_style.update(style_override)
+
+    # coerce node style values to strings, required by graphviz
+    node_style = {k: str(v) for k, v in node_style.items()}
+
+    return graph.node("N{}".format(node.pk), **node_style)
+
+
+def _add_graphviz_edge(graph, in_node, out_node, style=None):
+    """add graphviz edge between two nodes
+
+    :param graph: the graphviz.DiGraph to add the edge to
+    :param in_node: the head node
+    :param out_node: the tail node
+    :param style: the graphviz style (Default value = None)
+    :type style: dict or None
+
+    """
+    if style is None:
+        style = {}
+
+    # coerce node style values to strings
+    style = {k: str(v) for k, v in style.items()}
+
+    return graph.edge("N{}".format(in_node.pk), "N{}".format(out_node.pk), **style)
+
+
+class Graph(object):
+    """a class to create graphviz graphs of the AiiDA node provenance"""
+
+    # pylint: disable=useless-object-inheritance
+
+    def __init__(self,
+                 engine=None,
+                 graph_attr=None,
+                 global_node_style=None,
+                 global_edge_style=None,
+                 include_sublabels=True,
+                 link_styles=None,
+                 data_styles=None,
+                 process_styles=None,
+                 data_sublabels=None):
+        """a class to create graphviz graphs of the AiiDA node provenance
+
+        Nodes and edges, are cached, so that they are only created once
+
+        :param engine: the graphviz engine, e.g. dot, circo (Default value = None)
+        :type engine: str or None
+        :param graph_attr: attributes for the graphviz graph (Default value = None)
+        :type graph_attr: dict or None
+        :param global_node_style: styles which will be added to all nodes.
+            Note this will override any builtin attributes (Default value = None)
+        :type global_node_style: dict or None
+        :param global_edge_style: styles which will be added to all edges.
+            Note this will override any builtin attributes (Default value = None)
+        :type global_edge_style: dict or None
+        :param include_sublabels: if True, the note text will include node dependant sub-labels (Default value = True)
+        :type include_sublabels: bool
+        :param link_styles: callable mapping LinkType to graphviz style dict;
+            link_styles(link_type) -> dict (Default value = None)
+        :param data_styles: callable mapping data node to a graphviz style dict;
+            data_styles(node) -> dict (Default value = None)
+        :param process_styles: callable mapping process node to a graphviz style dict;
+            process_styles(node) -> dict (Default value = None)
+        :param data_sublabels: callable mapping data node to a sublabel (e.g. specifying some attribute values)
```

Comment:
Is there a reason to limit this to data nodes?

If somebody wants to overwrite the behavior, I can imagine they might also want to do it for other types of nodes (?)
If not, this parameter could be called `sublabel_fn`